drivers/spi/spi_ite_it8xxx2.c

```diff
@@ -0,0 +1,487 @@
+	}
+
+	spi_regs->INTSTS |= SPICMDQEND;
+	return err;
+
+}
+
+static int it8xxx2_transceive(const struct device *dev,
+			    const struct spi_config *config,
+			    const struct spi_buf_set *tx_bufs,
+			    const struct spi_buf_set *rx_bufs)
+{
+	struct spi_it8xxx2_data *data = dev->data;
+
+	size_t nbrx = 0;
+	size_t nbtx = 0;
+
+	const struct spi_buf *ptx = tx_bufs->buffers;
+	const struct spi_buf *prx = rx_bufs->buffers;
+
+	int err = 0;
+	int ret = 0;
+
+	ret = k_mutex_lock(&data->it8xxx2_mutex, K_FOREVER);
+
+	if (tx_bufs != NULL) {
+		nbtx = tx_bufs->count;
+	}
+
+	while (nbtx--) {
+		err = spi_it8xxx2_trans(dev, config, ptx, prx);
+		if (err != 0) {
+			LOG_ERR("[kern][spi_it8xxx2_trans] Returned value is not 0\r\n");
+		}
+	}
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = ret;
+	}
+
+	if (rx_bufs != NULL) {
+		nbrx = rx_bufs->count;
+	}
+
+	while (nbrx--) {
+		err = spi_it8xxx2_trans(dev, config, ptx, prx);
+		if (err != 0) {
+			LOG_ERR("[kern][spi_it8xxx2_trans] Returned value is not 0\r\n");
+		}
+	}
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = ret;
+	}
+
+	k_mutex_unlock(&data->it8xxx2_mutex);
+
+	return err;
+
+}
+
+static int it8xxx2_transceive_sync(const struct device *dev,
+				 const struct spi_config *config,
+				 const struct spi_buf_set *tx_bufs,
+				 const struct spi_buf_set *rx_bufs)
+{
+	return it8xxx2_transceive(dev, config, tx_bufs, rx_bufs);
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int it8xxx2_transceive_async(const struct device *dev,
+				  const struct spi_config *config,
+				  const struct spi_buf_set *tx_bufs,
+				  const struct spi_buf_set *rx_bufs,
+				  struct k_poll_signal *async)
+{
+	return it8xxx2_transceive(dev, config, tx_bufs, rx_bufs);
+}
+#endif
+
+static int it8xxx2_release(const struct device *dev,
+			 const struct spi_config *config)
+{
+	struct spi_it8xxx2_data *data = dev->data;
+
+	k_sem_give(&data->it8xxx2_sem);
+	return 0;
+}
+
+static void spi_it8xxx2_isr(const struct device *dev)
+{
+	const struct spi_it8xxx2_config *cfg = dev->config;
+	struct spi_it8xxx2_data *data = dev->data;
+	struct spi_it8xxx2_regs *const spi_regs =
+		(struct spi_it8xxx2_regs *)cfg->base_addr;
+
+	irq_disable(cfg->irq_no);
+
+	spi_regs->INTSTS |= SPICMDQEND;
+	ite_intc_isr_clear(cfg->irq_no);
+	spi_regs->SPICTRL5 &= ~(CH0SELCMDQ);
+	k_sem_give(&data->it8xxx2_sem);
+	chip_permit_idle();
+}
+
+static int spi_it8xxx2_init(const struct device *dev)
+{
+	const struct spi_it8xxx2_config *cfg = dev->config;
+	struct spi_it8xxx2_data *data = dev->data;
+
+	int status;
+
+	status = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+
+	if (status < 0) {
+		LOG_ERR("Failed to configure SPI pins\r\n");
+		return status;
+	}
+
+	/* Initialize Mutex */
+	k_mutex_init(&data->it8xxx2_mutex);
+	/* Initialize Semaphore */
+	k_sem_init(&data->it8xxx2_sem, 0, 1);
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
+	spi_it8xxx2_isr, DEVICE_DT_INST_GET(0), 0);
```

Comment:
Your code assumes SPI instance 0.  You can use this instead.
```suggestion
	IRQ_CONNECT(cfg->irq_no, cfg->irq_priority, spi_it8xxx2_isr, dev, 0);
```

Reply:
Got it, sir. Actually I think that we'd add the multiple-instance declaration back, but for supporting ASYNC option, the arguments spi_callback_t cb & void *userdata should be listed in the _transceive_async api, so I'm now trying to figure out how to replace original semaphore declaration and usage with helper functions in spi_context.h. So far this case is kinda pending temporarily. Thanks for great patience and guidnace, sir.